primitive-types/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+// except according to those terms.
+
+//! Primitive types shared by Substrate and Parity Ethereum.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(all(feature = "impl-serde", not(feature = "std")))]
+compile_error!("Feature \"impl-serde\" requires feature \"std\" to build.");
+
+#[cfg(all(feature = "impl-rlp", not(feature = "std")))]
+compile_error!("Feature \"impl-rlp\" requires feature \"std\" to build.");
+
+#[macro_use]
+extern crate uint;
+
+#[macro_use]
+extern crate fixed_hash;
+
+#[macro_use]
+extern crate crunchy;
+
+#[cfg(feature = "impl-serde")]
+#[macro_use]
+extern crate impl_serde;
+
+#[cfg(feature = "impl-codec")]
+#[macro_use]
+extern crate impl_codec;
+
+#[cfg(feature = "impl-rlp")]
+#[macro_use]
+extern crate impl_rlp;
+
+construct_uint!(U256, 4);
+#[cfg(feature = "impl-serde")] impl_uint_serde!(U256, 4);
+#[cfg(feature = "impl-codec")] impl_uint_codec!(U256, 4);
+#[cfg(feature = "impl-rlp")] impl_uint_rlp!(U256, 4);
+
+impl U256 {
+	/// Multiplies two 256-bit integers to produce full 512-bit integer
+	/// No overflow possible
+	#[inline(always)]
+	pub fn full_mul(self, other: U256) -> U512 {
+		U512(uint_full_mul_reg!(U256, 4, self, other))
+	}
+}
+
+impl From<U256> for U512 {
+	fn from(value: U256) -> U512 {
+		let U256(ref arr) = value;
+		let mut ret = [0; 8];
+		ret[0] = arr[0];
+		ret[1] = arr[1];
+		ret[2] = arr[2];
+		ret[3] = arr[3];
+		U512(ret)
+	}
+}
+
+impl From<U512> for U256 {
+	fn from(value: U512) -> U256 {
+		let U512(ref arr) = value;
+		if arr[4] | arr[5] | arr[6] | arr[7] != 0 {
+			panic!("From<U512> for U256: encountered overflow")
+		}
+		let mut ret = [0; 4];
+		ret[0] = arr[0];
+		ret[1] = arr[1];
+		ret[2] = arr[2];
+		ret[3] = arr[3];
+		U256(ret)
+	}
+}
+
+impl<'a> From<&'a U256> for U512 {
+	fn from(value: &'a U256) -> U512 {
+		let U256(ref arr) = *value;
+		let mut ret = [0; 8];
+		ret[0] = arr[0];
+		ret[1] = arr[1];
+		ret[2] = arr[2];
+		ret[3] = arr[3];
+		U512(ret)
+	}
+}
+
+impl<'a> From<&'a U512> for U256 {
+	fn from(value: &'a U512) -> U256 {
+		let U512(ref arr) = *value;
+		if arr[4] | arr[5] | arr[6] | arr[7] != 0 {
+			panic!("From<&U512> for U256: encountered overflow")
+		}
+		let mut ret = [0; 4];
+		ret[0] = arr[0];
+		ret[1] = arr[1];
+		ret[2] = arr[2];
+		ret[3] = arr[3];
+		U256(ret)
+	}
+}
+
+construct_uint!(U512, 8);
+#[cfg(feature = "impl-serde")] impl_uint_serde!(U512, 8);
+#[cfg(feature = "impl-codec")] impl_uint_codec!(U512, 8);
+#[cfg(feature = "impl-rlp")] impl_uint_rlp!(U512, 8);
+
+construct_fixed_hash! {
+	/// Fixed-size uninterpreted hash type with 20 bytes (160 bits) size.
+	pub struct H160(20);
+}
+#[cfg(feature = "impl-serde")] impl_fixed_hash_serde!(H160, 20);
+#[cfg(feature = "impl-codec")] impl_fixed_hash_codec!(H160, 20);
+#[cfg(feature = "impl-rlp")] impl_fixed_hash_rlp!(H160, 20);
+
+#[cfg(feature = "std")]
+impl From<u64> for H160 {
+	fn from(val: u64) -> Self {
+		Self::from_low_u64_be(val)
+	}
+}
+
+impl From<H160> for H256 {
```

Comment:
This should use `impl_fixed_hash_conversions!(H256, H160)` I think.

Reply:
We don't have that macro now and I don't think it's a good idea -- converting `H160` to `H256` is a special operation (that has agreement on Substrate and Parity Ethereum). We cannot apply it directly to any other types (otherwise, we can't answer issues like why is it zero-padded on the front? why not on the end or repeat the bytes?).

Reply:
I meant [this](https://github.com/paritytech/parity-common/blob/59d9672df718cf42316b4445c641aa977461c028/fixed-hash/src/hash.rs#L817); it does a static assert on the type size and zero-pads in front. I agree this is arbitrary and not something we invented recently, but it is there and it's been used for a long time so I'd argue it's the "standard by convention" in our codebase. As long as we document the behaviour I can live with it.

The reason I'm being difficult here and elsewhere is that I'm taking the POV of a dev that is unfamiliar with the code structure and approaches the code base for the first time: I'm afraid they will grow frustrated with how code is spread out in different crates with a hard-to-grasp logic as to what stuff is found where. If at all possible I'd like logic to be in `uint`/`fixed-hash`.

Reply:
Ah right. Fixed!